### PR TITLE
Add heading hierarchy checker notice

### LINKED ISSUE #10581: Add a headings hierarchy checker notice in the heading block sidebar
Previously, the headings hierarchy was placed also in the _Document_ sidebar, in the "Table of Contents" panel:

screen shot 2018-10-13 at 17 49 50

This panel has been removed so the only way to be aware of the incorrect headings level is to intentionally check the document outline tool in the top bar.

I'd like to propose to add a notice, even a small one, directly in the Heading block sidebar so users will be immediately informed the hierarchy is not correct _while_ they set the heading level. An option might be placing the notice under the buttons in the sidebar:

screenshot 103

### PR DESCRIPTION
**Pull request by Jackie6**
## Description
Fix #10581 

## How has this been tested?
- [X] local test
- [X] browser test

## Screenshots 
image

## Types of changes
Add a checker to check if the selected heading level is correct

## Checklist:
- [X] My code is tested.
- [X] My code follows the WordPress code style. 
- [X] My code follows the accessibility standards. 
- [X] My code has proper inline documentation. 
- [X] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by Jackie6** [comment]
@AmartyaU Hope you can help work on this issue!

**Comment by AmartyaU** [comment]
@Jackie6 and I talked about the content of the error messages, which has also been debated about in the parent issue. We also have a solution that details the exact heading level to choose in the error message, something like "This heading level is incorrect. Please use H1, or H2 instead."

Our question is if the error message should be short but vague or long and detailed?

**Comment by afercia** [comment]
Thanks for working on this! Works pretty well :)

Stepping in after @gziolo invitation to give some feedback on the message wording. Personally, I'd keep it as is because in the sidebar the message should be short, simple, and informative. "This heading level is incorrect." seems OK to me.

However, I'm wondering if the message should also point users to the "Content structure" tool in the top bar because:
- the information there is more detailed
- it would help improve discoverability of the "Content structure" tool

There are cases "Content structure" definitely gives more context. For example, a H1 in the post content _may_ be correct in some edge cases. Screenshots to illustrate the added context:

Screenshot 2019-04-10 at 12 15 19

Screenshot 2019-04-10 at 12 15 27

I know @joedolson has some thoughts on the message wording, I'd like to hear his opinion :)

**Comment by mapk** [comment]
> I'm wondering if the message should also point users to the "Content structure" tool in the top bar

@afercia how do you imagine this working? Would there be a link that opened this popover from the Block Inspector in relation to the messaging?

**Comment by Jackie6** [comment]
Hi, @afercia Thanks for comments. I could not imagine the edge cases where the H1 heading is correct, could you illustrate this a bit more? And I have the same question as @mapk, how would you expect to open the popover?

**Comment by joedolson** [comment]
The edge cases where an H1 would be correct would be pretty obscure, I think. I can think of at least one - if the theme omits the page title and the heading is the first block of content in the post. While that's unusual, I have seen cases, probably due to people wanting a different title used in lists of links than displayed on the page. 

I agree with the short wording; I think attempting to elaborate the wording in this context isn't beneficial. Giving a brief warning then providing access to the more detailed information would seem more user-friendly.

**Comment by afercia** [comment]
> The edge cases where an H1 would be correct would be pretty obscure, I think. I can think of at least one - if the theme omits the page title

Yep, also some post formats in some themes don't have a h1 heading, e.g. Twenty Thirteen "status" (if I remember correctly). Also, custom post types might not have a post title as well.

**Comment by afercia** [comment]
> how do you imagine this working? Would there be a link that opened this popover from the Block Inspector in relation to the messaging?

Wasn't actually thinking at a control to open the Content Structure. Interaction would be tricky: focus should be moved there, then moved back to the invoking context when closing, etc.

I was just thinking to specify in the text something like:
"Hey dear user, this heading level is not correct. You may want to check the Content Structure tool for more context". With some shorter wording, of course :)

File: packages/block-library/src/heading/heading-checker.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import { __ } from '@wordpress/i18n';
+import { Notice } from '@wordpress/components';
+import { useEffect } from '@wordpress/element';
+import { compose } from '@wordpress/compose';
+import { withSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { computeOutlineHeadings } from '../../../editor/src/components/document-outline/index.js';
```
Comment (gziolo) [inline packages/block-library/src/heading/heading-checker.js]: This won't work when published to npm. You can't reference private methods from another package this way. I guess this method should be copied over here instead as the first iteration.

File: packages/block-library/src/heading/heading-checker.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+	const msg = __( 'This heading level is incorrect. ' );
+
+	// For accessibility
+	useEffect( () => {
+		speak( __( 'This heading level is incorrect' ) );
+	}, [ selectedLevel ] );
+
+	return (
+		<div className="editor-heading-checker block-editor-heading-checker">
```
Comment (gziolo) [inline packages/block-library/src/heading/heading-checker.js]: Class names should follow the coding guidelines:

https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#css

In this case, it should be: `block-library-heading__heading-level-checker`.

It looks like contrast checker is part of `@wordpress/block-editor` package. However, in this case, it seems to be a better idea to leave it local as it's very specific to this block type.

File: packages/block-library/src/heading/heading-checker.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+			<Notice status="warning" isDismissible={ false }>
+				{ msg }
+			</Notice>
+		</div>
+	);
+};
+
+export default compose(
+	withSelect( ( select ) => {
+		const { getBlocks } = select( 'core/block-editor' );
```
Comment (gziolo) [inline packages/block-library/src/heading/heading-checker.js]: @jorgefilipecosta - is there a way to get all heading blocks only so we didn't have to use `computeOutlineHeadings` in the first place?

There is [getGlobalBlockCount](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/store/selectors.js#L270-L285), so something similar would be nice.

That way we could share the same selector in both places and remove this `computeOutlineHeadings` helper altogether.

File: packages/block-library/src/heading/heading-checker.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+import { useEffect } from '@wordpress/element';
+import { compose } from '@wordpress/compose';
+import { withSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { computeOutlineHeadings } from '../../../editor/src/components/document-outline/index.js';
+
+export const HeadingChecker = ( { blocks = [], selectedLevel, selectedHeadingId } ) => {
```
Comment (gziolo) [inline packages/block-library/src/heading/heading-checker.js]: How do you feel about renaming it to `HeadingLevelChecker` to better reflects what it validates?

The name of the file could be `heading/level-checker.js` in that case.

File: packages/block-library/src/heading/heading-checker.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	// Iterate headings to find prevHeadingLevel
+	let prevHeadingLevel = 1;
+	let i = 0;
+	for ( i = 0; i < headings.length; i++ ) {
+		if ( i >= 1 && headings[ i ].clientId === selectedHeadingId ) {
+			prevHeadingLevel = headings[ i - 1 ].level;
+		}
+	}
+
+	const isIncorrectLevel = ( selectedLevel === 1 || selectedLevel > prevHeadingLevel + 1 );
```
Comment (gziolo) [inline packages/block-library/src/heading/heading-checker.js]: `selectedLevel` can be read from the currently selected block. We don't need to pass it down.

**Review by gziolo — COMMENTED** [review]
Nice work, I left some feedback. It's mostly about code-related improvements. The functionality itself seems to work as expected but that you can polish with accesibility experts :)

**Comment by AmartyaU** [comment]
> I was just thinking to specify in the text something like:
> "Hey dear user, this heading level is not correct. You may want to check the Content Structure tool for more context". With some shorter wording, of course :)

I love this solution. It would be the most user-friendly without complicated code and recurring feedback. 

Something like  "This heading level is incorrect. Check the Content Structure tool for more context."

We can also link 'Content Structure' to a resource explaining what the tool is for people who have never used it before. Like you said, it would also increase discoverability this way!

**Comment by afercia** [comment]
> We can also link 'Content Structure tool' to a resource explaining what the tool is for people who have never used it before. Like you said, it would also increase discoverability this way!

Yep, good point. Worth noting Gutenberg and core will do this starting from 5.2 for the alt text description 🙂linking to [the W3C alt text tutorial](https://www.w3.org/WAI/tutorials/images/decision-tree/)

Screenshot 2019-04-11 at 15 49 23

We'd just need to identify a good resource (better if translated or translatable). I guess a link could also be added in a separate PR.

File: packages/block-library/src/heading/heading-checker.js (hunk 5 above)
Comment (Jackie6) [reply packages/block-library/src/heading/heading-checker.js]: Solve it

File: packages/block-library/src/heading/heading-checker.js (hunk 2 above)
Comment (Jackie6) [reply packages/block-library/src/heading/heading-checker.js]: Solve it

File: packages/block-library/src/heading/heading-checker.js (hunk 1 above)
Comment (Jackie6) [reply packages/block-library/src/heading/heading-checker.js]: Solve it. But I do not think it is a good coding style to copy-paste the existing function. Is it possible to just export the function and import it?

File: packages/block-library/src/heading/heading-checker.js (hunk 3 above)
Comment (Jackie6) [reply packages/block-library/src/heading/heading-checker.js]: In fact, `headings` has been computed in 
https://github.com/WordPress/gutenberg/blob/8722f49d3614a9c71d7a6a3af06eebec4c9bfa5b/packages/editor/src/components/document-outline/index.js#L67-L69
I am not sure whether we could avoid computing it again.

Comment (gziolo) [reply packages/block-library/src/heading/heading-checker.js]: We can if we create memoized shared selector, similar to how `getBlocks` is cached as of today. It might be a complicated task and that's why I asked @jorgefilipecosta for help.

File: packages/editor/src/components/document-outline/index.js [hunk 6]
```diff
@@ -48,7 +48,7 @@ const multipleH1Headings = [
  *
  * @return {Array} An array of heading blocks enhanced with the properties described above.
  */
-const computeOutlineHeadings = ( blocks = [], path = [] ) => {
+export const computeOutlineHeadings = ( blocks = [], path = [] ) => {
```
Comment (gziolo) [inline packages/editor/src/components/document-outline/index.js]: It no longer should be exported:
```suggestion
const computeOutlineHeadings = ( blocks = [], path = [] ) => {
```

File: packages/block-library/src/heading/style.native.scss [hunk 7]
```diff
@@ -2,3 +2,6 @@
 .blockText {
 	min-height: $min-height-heading;
 }
+.block-library-heading__heading-level-checker {
```
Comment (gziolo) [inline packages/block-library/src/heading/style.native.scss]: This style if really necessary should be applied to `style.scss` file which web version uses. This file is used by React Native app.

File: packages/block-library/src/heading/heading-checker.js (hunk 1 above)
Comment (gziolo) [reply packages/block-library/src/heading/heading-checker.js]: > But I do not think it is a good coding style to copy-paste the existing function.

Yes, 100% correct.

> Is it possible to just export the function and import it?

There is an alternative, see my other comment: https://github.com/WordPress/gutenberg/pull/14889#discussion_r274377528

**Comment by gziolo** [comment]
Screen Shot 2019-04-12 at 09 51 03

I compared with the document outline feature and it looks like H1 has some special handling. It should be replicated. There are 2 messages used:
https://github.com/WordPress/gutenberg/blob/c2dcb6b2fa264d51a15ff0aeedc53e17f69b9108/packages/editor/src/components/document-outline/index.js#L32
https://github.com/WordPress/gutenberg/blob/c2dcb6b2fa264d51a15ff0aeedc53e17f69b9108/packages/editor/src/components/document-outline/index.js#L36
I think they both fit in the sidebar as well.

**Comment by gziolo** [comment]
@jasmussen, @kjellr or @mapk - we are about to have 2 blocks using notices in the sidebar:

Screen Shot 2019-04-12 at 10 21 19
Screen Shot 2019-04-12 at 10 21 27

Can you advice how we can align margins and paddings for all notices so we don't have to worry about it when adding next one?

**Review by gziolo — CHANGES_REQUESTED** [review]
@Jackie6 - on your side the only things left to land this PR is to figure out how to replicate the behavior for H1 as explained in https://github.com/WordPress/gutenberg/pull/14889#issuecomment-482483002

**Comment by afercia** [comment]
@gziolo from your screenshot on https://github.com/WordPress/gutenberg/pull/14889#issuecomment-482483002 I see the column block is included in the document outline. That's totally new to me and a bit surprising. Can you please point me at the PR that added that and the related argumentations?
It feels wrong to me: the document outline should contain exclusively headings and no information related to layout.

File: packages/block-library/src/heading/heading-checker.js (hunk 3 above)
Comment (jorgefilipecosta) [reply packages/block-library/src/heading/heading-checker.js]: We could create a generic selector that returns the blocks of a given type, so it is not specific to this functionality.
But I'm not sure if it would fit our needs, computeOutlineHeadings does more than just returning the heading blocks, it computes, for example, the heading path required to show the path where a heading is nested on.
I think in this specific case just the selector would work, but in the checks, we do in the Content Structure we would still need something like computeOutlineHeadings to associate a path to each heading.
Exposing a selector that does what computeOutlineHeadings does (or similar functionality with paths) seems very specific to this functionality.
I guess a possible path here is just implementing level-checker.js as part of the editor module and reusing computeOutlineHeadings there.

**Comment by afercia** [comment]
Discussed during today's accessibility bug-scrub. Removing the "Needs Accessibility Feedback" label, as feedback has been given.

**Comment by kjellr** [comment]
> Can you advice how we can align margins and paddings for all notices so we don't have to worry about it when adding next one?

Thanks for the ping, @gziolo. We should add a `$grid-size * 3` margin to the bottom of the Notice component when it's in the sidebar, alongside [the spacing we use below base controls](https://github.com/WordPress/gutenberg/blob/d82a3bcde825bb23991c2790aad10844169d3f3c/packages/edit-post/src/components/sidebar/settings-sidebar/style.scss#L7). 

Ideally, we'd also adopt the `:last-child` rule that's there, but all of these Notice components appear to be only children of `div`s without a standardized class name. 😄 So for now, just using the `$grid-size * 3` bottom margin all the time is ok. 

Also — +1 on the idea of linking elsewhere within this message. Lots of users think heading levels are basically just font size changes, so they'll need some extra context there.

**Comment by sarahmonster** [comment]
"This heading level is incorrect" might feel a bit blunt, especially for the vast majority of users who likely use headings based on visual weight, rather than semantics. 

Instead of just telling them they're wrong, we should:
1. Provide a link to more information (as discussed here), _and_
2. Suggest what would be the "correct" choice here.

Suggesting a better choice will help users who don't have time to read technical documentation avoid frustration whilst still creating more accessible content.

So instead of:
> This heading level is incorrect.

This would be a more helpful message:
> This level doesn't match your document structure. Try an H3 or H2 instead. _Learn more_

**Comment by AmartyaU** [comment]
> Suggesting a better choice will help users who don't have time to read technical documentation avoid frustration whilst still creating more accessible content.

As far as I understand, users are not being linked to technical documentation in this case and are just being pointed to look in the Content Structure tool. Only those users who don't know about the Content Structure tool would have to look at the technical documentation about it.

**Comment by paaljoachim** [comment]
Can we get a status update to where this PR is currently at?

**Comment by ZebulanStanphill** [comment]
It looks like this PR has fallen stale/inactive. I've picked up the work in #22650. Thank you for getting this train moving, @Jackie6! It would have been a lot harder to get this feature working without your initial effort!
